### PR TITLE
EIP: Coinbase Calls

### PR DESCRIPTION
**Pull request by 3esmit**
## Simple Summary

Allow contracts to be called directly by `block.coinbase` (block validator), without a transaction.

## Abstract

_In proof-of-work blockchains, validators are known as miners._

The validator might want to execute functions directly, without having to sign a transaction. Some examples might be presenting a proof in a contract for an change which also benefits the validator. 

A notable example would be when a validator want to act as an [EIP-1077] Gas Relayer, incentivized to pick up fees from meta transactions. 
Without this change, they can do so by signing from any address a `gasPrice = 0` transaction with the gas relayed call. 
However this brings an overhead of a signed transaction by validator that does nothing, as `msg.sender` is never used, and there is no gas cost to EVM charge. 

This proposal makes possible to remove this unused ecrecover.

## Motivation

In order to reduce the overhead of calls that don't use `msg.sender` and are being called by validator with `tx.gasPrice = 0`. 

## Specification

The calls to be executed by `block.coinbase` would be included first at block, and would consume normally the gas of block, however they won't pay/cost gas, instead the call logic would pay the validator in other form. 

Would be valid to execute any calls without a transaction by the block coinbase, except when the validator call tries to read `msg.sender`, which would throw an invalid jump.

Calls included by the validator would have `tx.origin = block.coinbase` and `gas.price = 0` for the rest of call stack, the rest follows as normal calls.

## Rationale

TBD

## Backwards Compatibility

`tx.origin = block.coinbase` could cause some issues on bad designed contracts, such as using `tx.origin` to validate a signature, an analysis on how contracts use tx.origin might be useful to decide if this is a good choice.

## Test Cases

TBD

## Implementation

TBD

## Security Considerations

TBD

## Copyright

Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

[EIP-1077]: https://eips.ethereum.org/EIPS/eip-1077

### CODE REVIEW

File: EIPS/eip-2474.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+---
+eip: 2474
+title: Coinbase calls
+author: Ricardo Guilherme Schmidt (@3esmit)
+discussions-to: https://ethresear.ch/t/gas-abstraction-non-signed-block-validator-only-procedures/4388/2
+status: Draft
+type: Standards Track 
+category: Core
+created: 2020-01-19
+
```
Comment (MicahZoltu) [inline EIPS/eip-2474.md:10]:
```suggestion
```

File: EIPS/eip-2474.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+Allow contracts to be called directly by `block.coinbase` (block validator), without a transaction.
+
+## Abstract
+
+_In proof-of-work blockchains, validators are known as miners._
+
+The validator might want to execute functions directly, without having to sign a transaction. Some examples might be presenting a proof in a contract for an change which also benefits the validator. 
+
+A notable example would be when a validator want to act as an [EIP-1077] Gas Relayer, incentivized to pick up fees from meta transactions.
```
Comment (MicahZoltu) [inline EIPS/eip-2474.md:23]:
```suggestion
A notable example would be when a validator want to act as an [EIP-1077](./eip-1077.md) Gas Relayer, incentivized to pick up fees from meta transactions.
```

File: EIPS/eip-2474.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+## Security Considerations
+
+TBD
+
+## Copyright
+
+Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
+
+[EIP-1077]: https://eips.ethereum.org/EIPS/eip-1077
```
Comment (MicahZoltu) [inline EIPS/eip-2474.md:65]:
```suggestion
```

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Minor formatting adjustments, but otherwise looks good to merge.  @3esmit is this still desired given that EIP-1559 is likely to make it so miners have to pay some amount for gas?

**Comment by lightclient** [comment]
It looks like there was activity after the initial stale notice, so this shouldn't have been closed?

**Comment by MicahZoltu** [comment]
Hmm, good point...

**Comment by MicahZoltu** [comment]
@3esmit Waiting to merge this until I heard back from you on https://github.com/ethereum/EIPs/pull/2474#pullrequestreview-484760851

File: EIPS/eip-2474.md (hunk 3 above)
Comment (3esmit) [reply EIPS/eip-2474.md]: Done in 9e97881

**Comment by 3esmit** [comment]
@MicahZoltu This shouldn't relate to EIP-1559. 
The proposed change in EIP 2474 makes possible to miners to include transactions that don't need a signer, because the validation is done in the smart contract. 
If EIP-1559 makes miners have to pay some gas, this don't changes the fact that some transactions don't need a `msg.sender`. 

This EIP is likely being replaced by #2938

File: EIPS/eip-2474.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
+---
+eip: 2474
+title: Coinbase calls
+author: Ricardo Guilherme Schmidt (@3esmit)
+discussions-to: https://ethresear.ch/t/gas-abstraction-non-signed-block-validator-only-procedures/4388/2
+status: Draft
+type: Standards Track
```
Comment (MicahZoltu) [inline EIPS/eip-2474.md:7]:
```suggestion
type: Standards Track
```
To satisfy the validator bot.

**Review by MicahZoltu — APPROVED** [review]
@3esmit Would you prefer to close this PR in favor of Account Abstraction, or do you still want to pursue this separate from AA?  If you want to pursue it, accept the suggested code change so the bot green lights and then @mention me so I remember to come back and merge.  😄

**Comment by 3esmit** [comment]
@MicahZoltu Ok, you can merge. For now we can left it as draft. Once AA gets defined/final, then move this to "abandoned".

**Comment by lightclient** [comment]
@3esmit FYI in case you haven't seen, #2938 defines AA (with working code!).

**Comment by 3esmit** [comment]
> @3esmit FYI in case you haven't seen, #2938 defines AA (with working code!).

Awesome. Once 2938 gets final I will move this one to abandoned as it would not be necessary anymore.

**Pull request merged**